vault/sealunwrapper.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	if underTxn, ok := underlying.(physical.Transactional); ok {
+		return &transactionalSealUnwrapper{
+			sealUnwrapper: ret,
+			Transactional: underTxn,
+		}
+	}
+
+	return ret
+}
+
+var _ physical.Backend = (*sealUnwrapper)(nil)
+var _ physical.Transactional = (*transactionalSealUnwrapper)(nil)
+
+type sealUnwrapper struct {
+	underlying   physical.Backend
+	logger       log.Logger
+	locks        []*locksutil.LockEntry
+	allowUnwraps *uint32
+}
+
+// transactionalSealUnwrapper is a seal unwrapper that wraps a physical that is transactional
+type transactionalSealUnwrapper struct {
+	*sealUnwrapper
+	physical.Transactional
+}
+
+func (d *sealUnwrapper) Put(ctx context.Context, entry *physical.Entry) error {
+	if entry == nil {
+		return nil
+	}
+
+	locksutil.LockForKey(d.locks, entry.Key).Lock()
+	defer locksutil.LockForKey(d.locks, entry.Key).Unlock()
+
+	return d.underlying.Put(ctx, entry)
+}
+
+func (d *sealUnwrapper) Get(ctx context.Context, key string) (*physical.Entry, error) {
+	entry, err := d.underlying.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var performUnwrap bool
+	se := &physical.SealWrapEntry{}
+	// If the value ends in our canary value, try to decode the bytes.
+	eLen := len(entry.Value)
+	if eLen > 0 && entry.Value[eLen-1] == 's' {
+		if err := proto.Unmarshal(entry.Value[:eLen-1], se); err == nil {
+			// We unmarshaled successfully which means we need to store it as a
+			// non-proto message
+			performUnwrap = true
+		}
+	}
+	if !performUnwrap {
+		return entry, nil
+	}
+	// It's actually encrypted and we can't read it
+	if se.Wrapped {
+		return nil, fmt.Errorf("cannot decode sealwrapped storage entry %s", entry.Key)
+	}
+	if atomic.LoadUint32(d.allowUnwraps) != 1 {
+		return &physical.Entry{
+			Key:   entry.Key,
+			Value: se.Ciphertext,
+		}, nil
+	}
+
+	locksutil.LockForKey(d.locks, key).Lock()
+	defer locksutil.LockForKey(d.locks, key).Unlock()
+
+	// At this point we need to re-read and re-check
+	entry, err = d.underlying.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	performUnwrap = false
+	se = &physical.SealWrapEntry{}
+	// If the value ends in our canary value, try to decode the bytes.
+	eLen = len(entry.Value)
+	if eLen > 0 && entry.Value[eLen-1] == 's' {
+		// We ignore an error because the canary is not a guarantee; if it
+		// doesn't decode, proceed normally
+		if err := proto.Unmarshal(entry.Value[:eLen-1], se); err == nil {
+			// We unmarshaled successfully which means we need to store it as a
+			// non-proto message
+			performUnwrap = true
+		}
+	}
+	if !performUnwrap {
+		return entry, nil
+	}
+	if se.Wrapped {
+		return nil, fmt.Errorf("cannot decode sealwrapped storage entry %s", entry.Key)
+	}
+	if atomic.LoadUint32(d.allowUnwraps) != 1 {
+		return &physical.Entry{
+			Key:   entry.Key,
+			Value: se.Ciphertext,
+		}, nil
+	}
+
+	entry = &physical.Entry{
```

Comment:
You could create the entry above and just return it if not allowUnwraps.

Reply:
Done.